inference-engine/src/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -0,0 +1,683 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/op_conversions/einsum_decomposition.hpp"
+
+#include <memory>
+#include <ngraph/opsets/opset7.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/validation_util.hpp>
+#include <transformations/utils/utils.hpp>
+
+#include "itt.hpp"
+
+namespace {
+/// \brief      Check if the EinsumDecomposition transformation is applicable to a given Einsum.
+/// The transformation is applicable if input subscript does not have repeated labels and ellipsis.
+///
+/// \param      subscript          A subscript to check its format
+///
+/// \return     true - applicable, false - not applicable
+///
+bool is_subscript_applicable(const std::string& subscript) {
+    auto labels = ngraph::opset7::Einsum::extract_labels(subscript);
+    auto unique_labels = std::unordered_set<std::string>(labels.begin(), labels.end());
+    return std::find(labels.begin(), labels.end(), "...") == labels.end() && unique_labels.size() == labels.size();
+}
+
+/// \brief      Compute einsum_path for a given Einsum node meaning that the (pseudo-)optimal
+/// order of operands contraction in terms of performance and memory consumption
+///
+/// \param      einsum_node         An input Einsum node
+///
+/// \return     a vector of pairs with input indices assuming that the intermediate result is
+/// appended in the tail
+///
+std::vector<std::pair<size_t, size_t>> compute_einsum_path(std::shared_ptr<const ngraph::opset7::Einsum> einsum_node) {
+    // TODO: implement algorithm for finding (pseudo-)optimal einsum_path
+    std::vector<std::pair<size_t, size_t>> einsum_path;
+    const size_t num_inputs = einsum_node->get_input_size();
+    NGRAPH_CHECK(num_inputs > 0);
+    for (size_t input_ind = num_inputs - 1; input_ind > 0; --input_ind) {
+        einsum_path.push_back(std::make_pair(0, input_ind));
+    }
+    return einsum_path;
+}
+
+/// \brief      Check if the dimension with a given label is reduced. The dimension is reduced
+/// if the corresponding label is met in neither the output subscript nor the input subscripts
+/// excluding ones specified by a vector excluded_indices
+///
+/// \param      input_subscripts         The vector of the input subscripts
+/// \param      output_subscript         The output subscript
+/// \param      label_to_check           A label that corresponds to dimension to check
+/// \param      excluded_indices         A vector of input subscript indices to be excluded
+///
+/// \return     true - a dimension to reduce, false - otherwise
+///
+bool is_dimension_reduced(const std::vector<std::string>& input_subscripts, const std::string& output_subscript,
+    const std::string label_to_check, const std::vector<size_t>& excluded_indices) {
```

Comment:
```suggestion
    const std::string& label_to_check, const std::vector<size_t>& excluded_indices) {
```

Reply:
We agreed to fix it in the scope of another PR for reference Einsum Implementation